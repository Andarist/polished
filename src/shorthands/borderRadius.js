// @flow
import capitalizeString from '../internalHelpers/_capitalizeString'
import polish from '../validation/polish'

/**
 * A shorthand that accepts a value for side and a value for radius and applies the radius value to both corners of the side.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...borderRadius('top', '5px')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${borderRadius('top', '5px')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'borderTopRightRadius': '5px',
 *   'borderTopLeftRadius': '5px',
 * }
 */

function borderRadius(side: string, radius: string) {
  const uppercaseSide = capitalizeString(side)
  if (uppercaseSide === 'Top' || uppercaseSide === 'Bottom') {
    return {
      [`border${uppercaseSide}RightRadius`]: radius,
      [`border${uppercaseSide}LeftRadius`]: radius,
    }
  }

  if (uppercaseSide === 'Left' || uppercaseSide === 'Right') {
    return {
      [`borderTop${uppercaseSide}Radius`]: radius,
      [`borderBottom${uppercaseSide}Radius`]: radius,
    }
  }
  return {}
}

export default polish({
  modulePath: 'shorthands/borderRadius',
  types: [
    { key: 'side', type: 'cssDirection', required: true },
    { key: 'radius', type: 'cssMeasure', required: true },
  ],
})(borderRadius)
