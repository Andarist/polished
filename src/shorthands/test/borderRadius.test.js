// @flow
import borderRadius from '../borderRadius'

describe('borderRadius', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('returns the proper values for the top side', () => {
    expect(borderRadius('top', '5px')).toMatchSnapshot()
  })
  it('returns the proper values for the bottom side', () => {
    expect(borderRadius('bottom', '5px')).toMatchSnapshot()
  })
  it('returns the proper values for the right side', () => {
    expect(borderRadius('right', '5px')).toMatchSnapshot()
  })
  it('returns the proper values for the left side', () => {
    expect(borderRadius('left', '5px')).toMatchSnapshot()
  })

  it('should throw 2 errors when not passed any parameters', () => {
    borderRadius()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a direction', () => {
    borderRadius(null, '5px')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a radius', () => {
    borderRadius('top')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed an invalid direction', () => {
    borderRadius('all', '5px')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed an invalid radius', () => {
    borderRadius('top', 1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warrning when passed too many parameters', () => {
    borderRadius('top', '5px', '5px')
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
