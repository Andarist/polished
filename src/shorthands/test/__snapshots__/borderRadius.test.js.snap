exports[`borderRadius returns the proper values for the bottom side 1`] = `
Object {
  "borderBottomLeftRadius": "5px",
  "borderBottomRightRadius": "5px",
}
`;

exports[`borderRadius returns the proper values for the left side 1`] = `
Object {
  "borderBottomLeftRadius": "5px",
  "borderTopLeftRadius": "5px",
}
`;

exports[`borderRadius returns the proper values for the right side 1`] = `
Object {
  "borderBottomRightRadius": "5px",
  "borderTopRightRadius": "5px",
}
`;

exports[`borderRadius returns the proper values for the top side 1`] = `
Object {
  "borderTopLeftRadius": "5px",
  "borderTopRightRadius": "5px",
}
`;

exports[`borderRadius should throw 2 errors when not passed any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %crequires a %c1st parameter%c(%cside%c: %ccssDirection%c). However, you did not pass %cside%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %crequires a %c2nd parameter%c(%cradius%c: %ccssMeasure%c). However, you did not pass %cradius%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderRadius should throw a warrning when passed too many parameters 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- shorthands/borderRadius.js --%c

borderRadius %cexpects a maximum of %c2 parameters%c. However, you passed %c3 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderRadius should throw an error when not passed a direction 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %crequires a %c1st parameter%c(%cside%c: %ccssDirection%c). However, you did not pass %cside%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderRadius should throw an error when not passed a radius 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %crequires a %c2nd parameter%c(%cradius%c: %ccssMeasure%c). However, you did not pass %cradius%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderRadius should throw an error when passed an invalid direction 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %cexpects a %c1st parameter%c(%cside%c: %ccssDirection%c). However, you passed (%c\'all\'%c:%cstring%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderRadius should throw an error when passed an invalid radius 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderRadius.js --%c

borderRadius %cexpects a %c2nd parameter%c(%cradius%c: %ccssMeasure%c). However, you passed (%c\'1\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderRadius%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
